domain/secret/state/state.go

```diff
@@ -112,6 +114,189 @@ WHERE  label = $secretOwner.label`
+		if err != nil {
+			return errors.Trace(err)
+		}
+		dbSecretOwner := secretApplicationOwner{Label: label, ApplicationUUID: app_uuid}
+		err = tx.Query(ctx, checkExistsStmt, dbSecretOwner).Get(&dbSecretOwner)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Trace(domain.CoerceError(err))
+		}
+		if err == nil {
+			return fmt.Errorf("secret with label %q already exists for application %q or one of its units%w", label, appName, errors.Hide(secreterrors.SecretLabelAlreadyExists))
+		}
+		return nil
+	}
+}
+
+// CreateCharmUnitSecret creates a secret onwed by the specified unit,
+// returning an error satisfying [secreterrors.SecretAlreadyExists] if a secret
+// owned by the same unit with the same label already exists.
+// It also returns an error satisfying [uniterrors.NotFound] if
+// the unit does not exist.
+func (st State) CreateCharmUnitSecret(ctx context.Context, version int, uri *coresecrets.URI, unitName string, secret domainsecret.UpsertSecretParams) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	revisionUUID, err := uuid.NewUUID()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		dbSecretOwner := secretUnitOwner{SecretID: uri.ID, Label: secret.Label}
+
+		selectUnitUUID := `SELECT &M.uuid FROM unit WHERE unit_id=$M.unit_id`
+		selectUnitUUIDStmt, err := st.Prepare(selectUnitUUID, sqlair.M{})
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		result := sqlair.M{}
+		err = tx.Query(ctx, selectUnitUUIDStmt, sqlair.M{"unit_id": unitName}).Get(&result)
+		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return uniterrors.NotFound
+			} else {
+				return errors.Annotatef(err, "looking up unit UUID for %q", unitName)
+			}
+		}
+		dbSecretOwner.UnitUUID = result["uuid"].(string)
+
+		if err := st.createSecret(ctx, tx, version, uri, secret, revisionUUID, st.checkUnitSecretLabelExists(unitName, dbSecretOwner.UnitUUID)); err != nil {
+			return errors.Annotatef(err, "inserting secret records for secret %q", uri)
+		}
+
+		if err := st.upsertSecretUnitOwner(ctx, tx, dbSecretOwner); err != nil {
+			return errors.Annotatef(err, "inserting unit secret record for secret %q", uri)
+		}
+		return nil
+	})
+	return errors.Trace(domain.CoerceError(err))
+}
+
+// checkUnitSecretLabelExists returns function which checks if a charm unit secret with the given label already exists.
+func (st State) checkUnitSecretLabelExists(unitName, unit_uuid string) checkExistsFunc {
+	return func(ctx context.Context, tx *sqlair.TX, label string) error {
+		if label == "" {
+			return nil
+		}
+
+		// TODO(secrets) - we check using 2 queries, but should do in DDL
+		checkLabelExistsSQL := `
+SELECT secret_id AS &secretUnitOwner.secret_id
+FROM (
+SELECT secret_id
+FROM   secret_application_owner
+WHERE  label = $secretUnitOwner.label
+AND    application_uuid = (
```

Comment:
The clauses of both the queries in this union should be joins instead of sub-selects. You'll also avoid the double nesting in the second.